src-java/orientdb-migration/src/main/java/org/openkilda/FlowMirrorMigration.java

```diff
@@ -0,0 +1,91 @@
+/* Copyright 2023 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda;
+
+import com.orientechnologies.orient.core.db.ODatabaseSession;
+import com.orientechnologies.orient.core.db.OrientDB;
+import com.orientechnologies.orient.core.db.OrientDBConfig;
+import liquibase.change.custom.CustomTaskChange;
+import liquibase.change.custom.CustomTaskRollback;
+import liquibase.database.Database;
+import liquibase.exception.CustomChangeException;
+import liquibase.exception.RollbackImpossibleException;
+import liquibase.exception.SetupException;
+import liquibase.exception.ValidationErrors;
+import liquibase.resource.ResourceAccessor;
+
+public class FlowMirrorMigration implements CustomTaskChange, CustomTaskRollback {
+    private final String host = System.getenv("KILDA_ORIENTDB_HOST");
+    private final String dbName = System.getenv("KILDA_ORIENTDB_DB_NAME");
+    private final String user = System.getenv("KILDA_ORIENTDB_USER");
+    private final String password = System.getenv("KILDA_ORIENTDB_PASSWORD");
+    private ODatabaseSession db;
+
+    @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
+    private ResourceAccessor resourceAccessor;
+
+    @Override
+    public void execute(Database database) throws CustomChangeException {
+        String script = "BEGIN;\n";
+        script += "FOREACH ($path IN (SELECT FROM flow_mirror_path)) {\n";
+        script += "LET $fm = INSERT INTO flow_mirror (flow_mirror_id, forward_path_id, mirror_switch_id, "
+                + "egress_switch_id, status, egress_port, egress_outer_vlan, egress_inner_vlan) VALUES ($path.path_id, "
+                + "$path.path_id, $path.mirror_switch_id, $path.egress_switch_id, $path.status, $path.egress_port, "
+                + "$path.egress_outer_vlan, $path.egress_inner_vlan);\n";
+        script += "UPDATE EDGE owns SET in = $fm.@rid[0] UPSERT WHERE @rid = $path.in_owns[0];\n";
+        script += "CREATE EDGE owns FROM $fm TO $path;\n";
+        script += "};\n";
+        script += "COMMIT;\n";
```

Comment:
This migrate doesn't look like idempotent. If the data in DB is large it might fail in the middle and the only way to proceed is the rollback and try again. And the next attempt might fail again.

Maybe it is overengineering, but I'd think about making the script in a form of updating only not already updated edges and commit in batches. Then if it fails, only the last batch is not committed and you can restart the migrate to continue from where it stopped.

Reply:
Sure, please provide an example of code, which implements your ideas and I will add it.

Reply:
Also, just FYI this script runs within one pessimistic transaction.

Reply:
Maybe the first thing is to understand the boundaries. If the data set you are updating is very large and the operation is performed on a live DB, there could be all sorts of issues such as dead locks, slowness, connection time outs.
To address this, we can use batch operation. Most likely Orient DB supports it, but I am not an expert in this area, right now I don't have a receipt. Otherwise we can do it using plain SQL:
overall idea is that you select some subset, do operation, commit, select next subset, and so on. Semantically:
while there is data
begin
perform updates in a loop
commit
end
To select subset of input data you can use the select filtering what is already processed with limiting to some reasonable amount of data: SELECT FROM flow_mirror_path WHERE <a predicate for not processed paths*> LIMIT 1000
then you loop over that 1000 records, commit, select a new portion of 1000 records, update, commit, and so on Once there is no more data you exit the outer loop. If that scripts fails for whatever reason, you can start it again and only the last 1000 is not committed. So you'll resume the updating on the last portion.
* A predicate could be NOT EXIST (what you want to add) or a value in some column/property. In your case, I believe it is the records which don't have an edge you are adding.

Reply:
Let's wait for Sergey. I think he knows boundaries. Then we will be able to decide if we need some changes.
I don't expect a big amount of data.